Added LDAP Group Sync for Single LDAP Schema

stevekuznetsov opened the pull request:
@deads2k @liggitt 

Implementation in place, testing in place ~~but blocked from running by https://github.com/openshift/openldap/pull/6.~~ Implementation for following schema only:

1. Groups as first-class LDAP entries with an attribute containing a list of members:
  1. members listed with their DN in the attribute
  * members listed with some other UID in the attribute

stevekuznetsov commented on pkg/cmd/experimental/sync-groups/groupsyncer.go:
```diff
@@ -0,0 +1,284 @@
+				groupGetter: extractor,
+			}
+		}
+		extractor.groupNameMapper = nameMapper
```
When we wanted to break out a `GroupNameMapper` as a discrete object and allow us to substitute whatever for it, we run into this cyclical dependency where you can't create the `DataExtractor` without populating the `GroupNameMapper` but the `GroupNameMapper` needs  to have a pointer to a `groupGetter`, which is the `DataExtractor` itself.

deads2k commented on pkg/auth/authenticator/password/ldappassword/ldap.go:
```diff
@@ -85,48 +76,19 @@ func (a *Authenticator) AuthenticatePassword(username, password string) (user.In
-			c.ServerName = a.options.URL.Host
-		}
-		tlsConfig = &c
+	serverConfig := &ldaputil.LDAPServerConfig{
```
Why don't the `Options` contain an `LDAPServerConfig` directly?

deads2k commented on pkg/auth/authenticator/password/ldappassword/ldap.go:
```diff
@@ -85,48 +76,19 @@ func (a *Authenticator) AuthenticatePassword(username, password string) (user.In
 // connect returns an established ldap connection, or an error if the connection could not be made (or successfully upgraded to TLS)
 // if no error is returned, the caller is responsible for closing the connection
 func (a *Authenticator) connect() (*ldap.Conn, error) {
```
This method seems extraneous.

deads2k commented on pkg/auth/authenticator/password/ldappassword/ldap.go:
```diff
@@ -164,10 +126,10 @@ func (a *Authenticator) getIdentity(username, password string) (authapi.UserIden
-	attrs.Insert(a.options.AttributeName...)
-	attrs.Insert(a.options.AttributePreferredUsername...)
-	attrs.Insert(a.options.AttributeID...)
+	attrs.Insert(a.options.IdentityAttributeMapping.EmailAttribute...)
```
Is it more reasonable to have a `RequestedAttributes` method on `LDAPUserToIdentityMapping` that returns a `util.StringSet` of these?

deads2k replied:
This information is very tightly tied to the `IdentityFactory`, refactor to make the linkage very clear.

stevekuznetsov replied:
I wasn't sure if `Config` type objects were OK to put methods on - seems like most that we have do not have methods. It's unclear what refactor you're envisioning here, please elaborate.

deads2k replied:
> I wasn't sure if Config type objects were OK to put methods on - seems like most that we have do not have methods but that's definitely something I want to do. It's unclear what refactor you're envisioning here, please elaborate.

hmmm... I suspect you may hate users more than me.....

deads2k replied:
> I wasn't sure if Config type objects were OK to put methods on - seems like most that we have do not have methods but that's definitely something I want to do. It's unclear what refactor you're envisioning here, please elaborate.

Scheme'd objects do not have methods on them, but other valueobjects are allowed to have methods.

deads2k commented on pkg/auth/authenticator/password/ldappassword/ldap.go:
```diff
@@ -225,41 +187,13 @@ func (a *Authenticator) getIdentity(username, password string) (authapi.UserIden
-	uid := getAttributeValue(entry, a.options.AttributeID)
-	if uid == "" {
-		return nil, false, fmt.Errorf("Could not retrieve a non-empty value from %v attributes for dn=%q", a.options.AttributeID, entry.DN)
+	identityFactory := ldaputil.DefaultLDAPUserIdentityFactory{
```
This factory is stateless, right?  Why not attach it to the `Authenticator`?

deads2k commented on pkg/auth/authenticator/password/ldappassword/util/getattribute.go:
```diff
@@ -0,0 +1,26 @@
+	"github.com/go-ldap/ldap"
+)
+
+func GetAttributeValue(entry *ldap.Entry, attributes []string) string {
```
godoc

deads2k commented on pkg/auth/authenticator/password/ldappassword/util/identityfactory.go:
```diff
@@ -0,0 +1,60 @@
+// DefaultLDAPUserIdentityFactory creates Identities for LDAP user entries using an LDAPUserToIdentityMapping
+type DefaultLDAPUserIdentityFactory struct {
+	ProviderName string
+	Mapping      *LDAPUserToIdentityMapping
```
Why a pointer?  It's a small object that we probably don't want to mutate in a non-obvious part of the code.

deads2k commented on pkg/auth/authenticator/password/ldappassword/util/server.go:
```diff
@@ -0,0 +1,73 @@
+)
+
+// LDAPServerConfig holds information for connecting to an LDAP server
+type LDAPServerConfig struct {
```
Give me a way to construct this from a `LDAPURL`

stevekuznetsov replied:
I'm not convinced that this is possible. Things `LDAPServerConfig` has that `LDAPURL` does not:

* `insecure`
* `TLSConfig`
* `TimeLimit`

deads2k replied:
> I'm not convinced that this is possible. Things LDAPServerConfig has that LDAPURL does not:

Default them and/or take params.  Probably different flavors for the different security options.

stevekuznetsov replied:
I see. Your original comment made it sound like you wanted the only input to be the `LDAPURL`. That's reasonable.

deads2k commented on pkg/auth/authenticator/password/ldappassword/util/identityfactory.go:
```diff
@@ -0,0 +1,60 @@
+package util
```
util packages are nasty things.  How about `ldaputil`?

deads2k commented on pkg/auth/userregistry/identitymapper/provision.go:
```diff
@@ -189,7 +189,7 @@ func (p *provisioningIdentityMapper) getMapping(ctx kapi.Context, identity *user
 }
 
-func getPreferredUserName(identity *userapi.Identity) string {
+func GetPreferredUserName(identity *userapi.Identity) string {
```
Why make this public?

deads2k commented on pkg/cmd/experimental/syncgroups/config/types.go:
```diff
@@ -0,0 +1,204 @@
+	// LDAPSyncTypeFirstClassUserss defines an LDAP schema where users are stored as first-class LDAP
+	// entries with an attribute containing a list of groups they are a member of, with groups listed
+	// by a unique name, with no additional group metadata in the LDAP server.
+	LDAPSchemaTypeFirstClassUserss LDAPSchemaType = "FirstClassUserss"
```
double s typo

deads2k commented on pkg/cmd/experimental/syncgroups/config/types.go:
```diff
@@ -0,0 +1,204 @@
+)
+
+// LDAPSyncConfig contains all of the configuration information for an LDAP group sync job.
+type LDAPSyncConfig struct {
```
You need a new scheme with internals and externals.  See the `pkg/cmd/server/api` for an example.  Also, open an issue to make a fuzzer unit test that can easily accept arbitrary schemes.

deads2k replied:
Is the intent of this object to be serialized or built from a serialized config?

stevekuznetsov replied:
This object is built. The following fields are added:

* GroupClient
* Source
* Scope
* WhitelistContents

deads2k commented on pkg/cmd/experimental/syncgroups/config/types.go:
```diff
@@ -0,0 +1,204 @@
+	WhitelistContents []string // added to loaded config from cli input
+
+	// Type determines which LDAP server schema exists for the sync
+	Type LDAPSchemaType
```
I don't think I like this.  I'd rather split into different types based on the `LDAPSchemaType` rather than have one object that contains it all.

stevekuznetsov replied:
This is something I borrowed from our `buildapi` - it was a very common thing there. Pros/cons? Clearly having `<nil>` parts of the config object isn't the best, but otherwise we will have three objects that are exactly the same save this one part.

deads2k commented on pkg/cmd/experimental/syncgroups/groupsyncer.go:
```diff
@@ -0,0 +1,260 @@
+
+// GroupSyncer runs a Sync job on Groups
+type GroupSyncer interface {
+	Sync() (err error)
```
I haven't read all the way down, but this definitely feels like a case where we'd want `[]error`

deads2k commented on pkg/cmd/experimental/syncgroups/groupsyncer.go:
```diff
@@ -0,0 +1,260 @@
+		// get membership data
+		memberEntries, err := s.groupMemberExtractor.ExtractMembers(ldapGroupUID)
+		if err != nil {
+			return err
```
Rather than dying on a failure, make an error type that includes the `ldapGroupUID` and the error to return a slice of failures.

deads2k commented on pkg/cmd/experimental/syncgroups/groupsyncer.go:
```diff
@@ -0,0 +1,260 @@
+		return nil, err
+	}
+
+	if err != nil && kerrs.IsNotFound(err) {
```
don't need the nil check

stevekuznetsov replied:
We should update our `UserIdentityMapper` code as well - this is taken from there.

deads2k commented on pkg/cmd/experimental/syncgroups/groupsyncer.go:
```diff
@@ -0,0 +1,260 @@
+		}
+		return nil, fmt.Errorf("no group found with name: %s", groupName)
+	}
+	s.addAnnotations(group, ldapGroupUID)
```
No.  This method should not mutate the annotations, only get the Group itself.

deads2k commented on pkg/cmd/experimental/syncgroups/groupsyncer.go:
```diff
@@ -0,0 +1,260 @@
+	group.Annotations[syncconfig.LDAPSyncTime] = strconv.FormatInt(time.Now().Unix(), 10)
+}
+
+func (s *LDAPGroupSyncer) updateGroup(group *ouserapi.Group, usernames []string) error {
```
Have this method update the annotations.

deads2k commented on pkg/cmd/experimental/syncgroups/groupsyncer.go:
```diff
@@ -0,0 +1,260 @@
+}
+
+// Sync allows the LDAPGroupSyncer to be a GroupSyncer
+func (s *LDAPGroupSyncer) Sync() error {
```
This is a very pretty looking method.  Why can't all your code be like this? :)

stevekuznetsov replied:
To be fair what you've reviewed up to this point is mostly @liggitt's code :)

deads2k commented on pkg/cmd/experimental/syncgroups/groupsyncer.go:
```diff
@@ -0,0 +1,260 @@
+				return nil, err
+			}
+
+			s.addAnnotations(createdGroup, ldapGroupUID)
```
don't add annotations here

deads2k commented on pkg/cmd/experimental/syncgroups/groupsyncer.go:
```diff
@@ -0,0 +1,260 @@
+		if s.config.ProvisionGroups {
+			// Try to create a group with the available name
+			desiredGroup := &ouserapi.Group{
+				ObjectMeta: kapi.ObjectMeta{
```
Add the correct label here

deads2k replied:
> Add the correct label here

Ah, I see, it's annotation.  Only add the `ldapGroupUID` annotation.

deads2k replied:
You ought to have an LDAP server label here.

deads2k commented on pkg/cmd/experimental/syncgroups/groupsyncer.go:
```diff
@@ -0,0 +1,260 @@
+		return nil, fmt.Errorf("no group found with name: %s", groupName)
+	}
+	s.addAnnotations(group, ldapGroupUID)
+	return group, nil
```
check the ldap group UID annotation before returning.

deads2k commented on pkg/cmd/experimental/syncgroups/groupsyncer.go:
```diff
@@ -0,0 +1,260 @@
+	return err
+}
+
+func New(config syncconfig.LDAPSyncConfig) (*LDAPGroupSyncer, error) {
```
No, this is insane to read.  You need to decompose how the pieces are built.  This `New` method should contain the interfaces used by `LDAPGroupSyncer`.  Those individual interfaces should be built with much tighter argument lists.    In fact, I'd say you should probably not have a single exposed method that accepts enough information to build more than one kind of `GroupLister` as a for instance.

stevekuznetsov commented:
@deads2k I'm happy with what I've done

stevekuznetsov commented:
@ncdc Timeout panics [here](https://travis-ci.org/openshift/origin/jobs/74957049#L235):
```
panic: test timed out after 1m0s
goroutine 52 [running]:
testing.func·008()
	/usr/local/go/src/testing/testing.go:681 +0x168
created by time.goFunc
	/usr/local/go/src/time/sleep.go:129 +0x59
goroutine 1 [chan receive]:
testing.RunTests(0xbfb318, 0xf996a0, 0x8, 0x8, 0xc208095b01)
	/usr/local/go/src/testing/testing.go:556 +0xdc3
testing.(*M).Run(0xc20809b2c0, 0x19405e0)
	/usr/local/go/src/testing/testing.go:485 +0xe1
main.main()
	github.com/openshift/origin/pkg/dockerregistry/_test/_testmain.go:112 +0x382
```

ncdc commented:
@stevekuznetsov this is a flake caused by a problem retrieving data from the Docker Hub. We have a few unit tests that actually depend on external systems like GitHub or the Docker Hub to be online. See #3624.

ncdc commented:
I can't kick travis but perhaps @smarterclayton or @liggitt could

stevekuznetsov commented:
@ncdc testing this isn't time-sensitive, just wanted to let know know what I was seeing!

smarterclayton commented:
Closing and reopening sometimes kicks travis

deads2k commented on pkg/auth/authenticator/password/ldappassword/ldap.go:
```diff
@@ -15,41 +12,33 @@ import (
-	providerName string
-	options      Options
-	mapper       authapi.UserIdentityMapper
+	providerName    string
```
dead?

stevekuznetsov replied:
No, this is the authenticator code for `ldappassword` - still needs the provider name.

deads2k commented on pkg/auth/authenticator/password/ldappassword/util/identityfactory.go:
```diff
@@ -0,0 +1,70 @@
+package ldaputil
```
change foldername to match package name.

liggitt commented on pkg/cmd/server/origin/auth.go:
```diff
@@ -459,15 +460,16 @@ func (c *AuthConfig) getPasswordAuthenticator(identityProvider configapi.Identit
 			URL:          url,
-			Insecure:     provider.Insecure,
-			TLSConfig:    tlsConfig,
+			ServerConfig: ldaputil.New(url, provider.Insecure, tlsConfig, 0), // make time limit configurable?
```
`ldaputil.New` gives no indication of what it will return... should be `ServerConfig` or something, though that feels weird in a util package

liggitt commented on pkg/auth/authenticator/password/ldappassword/util/server.go:
```diff
@@ -0,0 +1,84 @@
+)
+
+// New returns a new LDAPServerConfig
+func New(url LDAPURL, insecure bool, tlsConfig *tls.Config, timeLimit int) LDAPServerConfig {
```
timeLimit probably shouldn't be server-wide... that's a per-request thing

stevekuznetsov replied:
Would you reasonably expect to be setting a timeout differently for different data fetch requests?

liggitt replied:
doing group syncs, sure. a query that can return 10,000 items I'd probably let run a lot longer than a single user entry lookup

stevekuznetsov replied:
:+1:

liggitt commented on pkg/auth/authenticator/password/ldappassword/util/server.go:
```diff
@@ -0,0 +1,84 @@
+}
+
+// LDAPServerConfig holds information for connecting to an LDAP server
+type LDAPServerConfig struct {
```
also, the name is weird. is this actually a client config?

deads2k replied:
Confirm this object was a straight move so I don't have to re-read?

stevekuznetsov replied:
Straight move

deads2k commented on pkg/auth/authenticator/password/ldappassword/util/identityfactory.go:
```diff
@@ -0,0 +1,70 @@
+
+// LDAPUserToIdentityMapping holds LDAP attributes that map to OpenShift Identity fields in oder to
+// create a deterministic mapping from LDAP user entries to OpenShift Identities
+type LDAPUserToIdentityMapping struct {
```
I think I'd like to take `Mapping` out of the name so I don't get confused with other `IdentityMappings`.  `LDAPEntryAttributes` perhaps?

stevekuznetsov replied:
I don't like `LDAPEntryAttributes` because this only makes sense for a *user* entry. I also don't like `LDAPEntryAttributes` because, while this is just a collection of attribute keys,  I want the purpose of the collection to be obvious from the name.

If you don't like the name collision with other `IdentityMapping` objects, what about `LDAPUserEntryAttributeMapping`?

deads2k commented on pkg/auth/authenticator/password/ldappassword/util/identityfactory.go:
```diff
@@ -0,0 +1,70 @@
+	id := authapi.NewDefaultUserIdentityInfo(f.ProviderName, uid)
+
+	// Add optional extra attributes if present
+	for k, attrs := range map[string][]string{
```
I wasn't a fan the first time I review this and I'm still not a fan.  Can you turn thes into`.Email(*ldap.Entry) string` sort of methods the `LDAPUserToIdentityMapping` object.

stevekuznetsov replied:
Sure, that's definitely more readable.

deads2k commented on pkg/auth/authenticator/password/ldappassword/ldap.go:
```diff
@@ -15,41 +12,33 @@ import (
-	TLSConfig *tls.Config
+	URL ldaputil.LDAPURL
+	// ServerConfig holds information about the LDAP server
+	ServerConfig ldaputil.LDAPServerConfig
```
Looks to me like this refactor (and what it depends on) is logically distinct from your other changes.  Can you refactor the commits and pull this into a pull by itself?  I think I like it and I think we'll have a lot fewer comments than in the main body of this pull.

deads2k commented on pkg/cmd/experimental/syncgroups/groupsyncer.go:
```diff
@@ -0,0 +1,157 @@
+	group.Users = usernames
+
+	// add LDAP-sync-specific annotations
+	s.addAnnotations(group, ldapGroupUID)
```
just inline here.

deads2k commented on pkg/cmd/experimental/syncgroups/config/types.go:
```diff
@@ -0,0 +1,148 @@
+
+	// ProvisionGroups determines whether or not the GroupSyncer will create new OpenShift Groups if
+	// they do not yet exist or ignore those sync requests that do not map to existing OpenShift Groups
+	ProvisionGroups bool
```
Eliminate this.  If they want this behavior, they should sync based on the openshift groups.

stevekuznetsov replied:
I see this is also redundant with a whitelist. That's a better way of getting the same result.

deads2k commented on pkg/cmd/experimental/syncgroups/grouplister.go:
```diff
@@ -0,0 +1,109 @@
+)
+
+// NewAllOpenShiftGroupLister returns a new AllLocalGroupLister
+func NewAllOpenShiftGroupLister(ldapSyncConfig config.LDAPSyncCommonConfig) interfaces.LDAPGroupLister {
```
Accept the arguments you need.  In this case, the `dapSyncConfig.CLIOptions.GroupClient` and the `ldapSyncConfig.ServerConfig`

deads2k commented on pkg/cmd/experimental/syncgroups/config/types.go:
```diff
@@ -0,0 +1,148 @@
+)
+
+// LDAPSyncCommonConfig contains all of the config that is common to all LDAP schemas
+type LDAPSyncCommonConfig struct {
```
This feels like an object that should not be.

deads2k replied:
Delete

stevekuznetsov replied:
Please explain why.

deads2k replied:
> Please explain why.

This abstraction is in the wrong place.  You've created a single uber-object to hold sufficient information to build every piece in this entire flow, but if the constructing code has enough information to build this object and enough insight to organize it all properly, it had enough information organized well enough to construct the sync job itself.  That makes this object an extraneous abstraction that isn't simplifying your code and introduces unusual dependencies and information leaks.

deads2k replied:
> This abstraction is in the wrong place. You've created a single uber-object to hold sufficient information to build every piece in this entire flow, but if the constructing code has enough information to build this object and enough insight to organize it all properly, it had enough information organized well enough to construct the sync job itself. That makes this object an extraneous abstraction that isn't simplifying your code and introduces unusual dependencies and information leaks.

More succinctly: "it smells bad"

stevekuznetsov replied:
But this object *doesn't* have that information. Look at the schema-specific `LDAPSyncConfig` structs - there's a reason I put the other two schemas in this pull, so it was obvious where I was going. This is just all of the config that will be the same for every schema. The utility of this object is so that I don't repeat myself on all of those pieces for each schema config.

deads2k commented on pkg/cmd/experimental/syncgroups/grouplister.go:
```diff
@@ -0,0 +1,109 @@
+
+// NewWhitelistOpenShiftGroupLister returns a new LocalGroupLister
+func NewWhitelistOpenShiftGroupLister(whitelist []string, client osclient.GroupInterface) (interfaces.LDAPGroupLister, error) {
+	groups, err := config.GetOpenShiftGroups(whitelist, client)
```
Don't do this here, do it in the `ListGroups` method.

deads2k commented on pkg/cmd/experimental/syncgroups/grouplister.go:
```diff
@@ -0,0 +1,109 @@
+	if err != nil {
+		return nil, err
+	}
+	return &LocalGroupLister{
```
Use a type that saves the whitelist and client

deads2k commented on pkg/cmd/experimental/syncgroups/groupsyncer.go:
```diff
@@ -0,0 +1,157 @@
+	// Allows the Syncer to search for OpenShift Groups
+	GroupClient client.GroupInterface
+	// Config stores all sync configuration
+	Config syncconfig.LDAPSyncCommonConfig
```
This shouldn't be here.  Have just a string for the label value.

deads2k commented on pkg/cmd/experimental/syncgroups/schema1/memberextractor.go:
```diff
@@ -0,0 +1,172 @@
+)
+
+// New builds a new GroupMemberExtractor using a schema-appropriate config
+func New(config LDAPSyncConfig) GroupMemberExtractor {
```
don't collide names with packages.

deads2k replied:
Don't use this config object.  I'm 99% sure you've build the wrong object.  Let me see the individual required objects flow into `New*` methods and we'll build back out from there.

deads2k replied:
Better name.  Looks like a `NewMemberExtractor` to me.

stevekuznetsov replied:
Need your opinion on `LDAPInterface` for a name - one thing it *can* do is extract members, but it also lists groups, gets a group or member entry for a name, etc.

deads2k replied:
> Need your opinion on LDAPInterface for a name - one thing it can do is extract members, but it also lists groups, gets a group or member entry for a name, etc.

Doesn't seem immediately bad.  Try it and I'll see how I like it on the re-review

deads2k commented on pkg/cmd/experimental/syncgroups/schema1/memberextractor.go:
```diff
@@ -0,0 +1,172 @@
+
+// GroupMemberExtractor extracts the member list of an LDAP group entry from an LDAP server
+// with first-class LDAP entries for groups
+type GroupMemberExtractor struct {
```
This is satisfying several interfaces.  Tell me which ones in the godoc.  Also, to avoid bugs as you refactor (heavily), try somethign like `var _ interfaces.GroupExtractor = &GroupMemberExtractor{}`

deads2k commented on pkg/cmd/experimental/syncgroups/config/types.go:
```diff
@@ -0,0 +1,148 @@
+	// extracted as annotations on the OpenShift Group mapped to that entry. Each attribute is mapped
+	// to an annotation key in this parameter. The first attribute with a non-empty value is used.
+	// Namespacing annotations is advised to reduce name collision.
+	AnnotationAttributes map[string][]string
```
delete this until we need it.

deads2k commented on pkg/cmd/experimental/syncgroups/groupnamemapper.go:
```diff
@@ -0,0 +1,59 @@
+}
+
+// NewDynamicGroupNameMapper returns a new EntryAttributeLDAPGroupNameMapper
+func NewDynamicGroupNameMapper(mapping config.GroupFieldAttributeMapping,
```
`NewAttributeGroupNameMapper`

deads2k commented on pkg/cmd/experimental/syncgroups/schema1/newsyncer.go:
```diff
@@ -0,0 +1,58 @@
+
+// NewGroupSyncer creates a GroupSyncer for a schema-specific config written for an LDAP schema with
+// first-class group entries that hold membership lists as attributes
+func NewGroupSyncer(ldapSyncConfig LDAPSyncConfig) (syncgroups.GroupSyncer, error) {
```
Move this into a stubbed out command package.  Flag it with a TODO to be deleted.  Have it accept the absolute minimum arguments you need to construct a job that runs.  Default absolutely everything you can.  It should pretty much be `ldapurl string`.

That will help you start to factor this more effectively.  You are still conflating user config about making choice with objects used to configure the runtime bits.

deads2k commented on pkg/cmd/experimental/syncgroups/schema2/config.go:
```diff
@@ -0,0 +1,16 @@
+package schema2
```
Move this entire package out of this pull.  I'm not going to review it until after schema1 is done.

deads2k commented on pkg/cmd/experimental/syncgroups/schema3/config.go:
```diff
@@ -0,0 +1,28 @@
+package schema3
```
Move this entire package out of this pull.  I'm not going to review it until after schema1 is done.

deads2k commented on pkg/cmd/experimental/syncgroups/sync-groups.go:
```diff
@@ -0,0 +1,174 @@
+	// ---groupsyncscope
+	// ------if it's not an -all, inject whitelist
+	// create syncer
+	return nil, nil
```
Build the syncer right here.  You should be able to take your options, build the intermediate object directly in this method and compose your syncer in place right here.  Completely delete your existing `LDAPSyncCommonConfig` and anythign else I marked up there.

deads2k commented on pkg/cmd/experimental/syncgroups/config/types.go:
```diff
@@ -0,0 +1,148 @@
+}
+
+// LDAPSyncCommandLineOptions contains all of the options and references that are derived from the user input
+type LDAPSyncCommandLineOptions struct {
```
Delete

stevekuznetsov replied:
Why? I realize at this moment this is analogous to the `Options` struct in the cli file - one should remain, yes?

deads2k replied:
> Why? I realize at this moment this is analogous to the Options struct in the cli file - one should remain, yes?

This is abstraction in the wrong location.  The name is an immediate red mark (command line options shouldn't be represented beyond the first layer in a command.  The content of the object looks like amalgamation of information that is meant to satisfy knowledge requirements for multiple potential implementations of an interface.  Also a no-no beyond the first two layers or so.  One layer to parse the input, organize it, and abstract the command line flags themselves.  One layer to convert that information into what actually needs to be run.  Last layer acts on the knowledge to do the work.  That's usually sufficient.

deads2k commented on pkg/cmd/experimental/syncgroups/config/types.go:
```diff
@@ -0,0 +1,148 @@
+}
+
+// GroupSyncScope determines the scope of the group sync operation
+type GroupSyncScope string
```
Delete

stevekuznetsov replied:
Why?

deads2k replied:
> Why?

Its in the wrong spot. Only the CLI should be caring about that.

deads2k commented on pkg/cmd/experimental/syncgroups/config/types.go:
```diff
@@ -0,0 +1,148 @@
+}
+
+// GroupSyncSource determines the source of the groups to be synced
+type GroupSyncSource string
```
Delete

stevekuznetsov replied:
Why?

deads2k commented on pkg/cmd/experimental/syncgroups/config/types.go:
```diff
@@ -0,0 +1,148 @@
+
+// LDAPCommonQueryConfig holds config for an LDAP query that retrieves all relevant entries from the
+// LDAP server - all users, all groups, etc.
+type LDAPCommonQueryConfig struct {
```
Don't use the word config in this name.  What is this?  `LDAPQueryTemplate`?

deads2k replied:
`LDAPQueryScope`?

deads2k commented on pkg/cmd/experimental/syncgroups/config/types.go:
```diff
@@ -0,0 +1,148 @@
+
+// LDAPSpecificQueryConfig holds config for a specific LDAP query that retrieves one specific relevant
+// entry from the LDAP server - one user, group, list of members that represent a group, etc.
+type LDAPSpecificQueryConfig struct {
```
`LDAPQueryTemplate`

deads2k commented on pkg/cmd/experimental/syncgroups/config/types.go:
```diff
@@ -0,0 +1,148 @@
+
+// LDAPSpecificQueryFromAttributeConfig holds config for a specific LDAP query where the specfic filter
+// is string-formatted with the value of an attribute on a first-class LDAP entry
+type LDAPSpecificQueryFromAttributeConfig struct {
```
`LDAPFilteredQueryTemplate

deads2k commented on pkg/cmd/experimental/syncgroups/config/types.go:
```diff
@@ -0,0 +1,148 @@
+package config
```
These aren't config objects, take them out of the config package.

deads2k replied:
They fit better in ldaputil.  Be sure to keep them separate from the other pull/commit.

stevekuznetsov replied:
I disagree here, I don't think these query templates and annotation/label keys (only used by the sync, not used at all by the ldap password auth system) should remain in this place. I've changed the package they're in but I have not moved them to the ldappassword package.

deads2k commented on pkg/cmd/experimental/syncgroups/schema1/ldapinterface.go:
```diff
@@ -0,0 +1,183 @@
+)
+
+// NewLDAPInterface builds a new LDAPInterface using a schema-appropriate config
+func NewLDAPInterface(config LDAPSyncConfig) LDAPInterface {
```
Don't accept `LDAPSyncConfig` here.  Take the four args you need.

deads2k commented on pkg/cmd/experimental/syncgroups/sync-groups.go:
```diff
@@ -0,0 +1,188 @@
+	return knownSources.Has(string(source))
+}
+
+type SyncGroupsOptions struct {
```
Have options contain the values for making the actual `LDAPGroupSyncer`.  Fill them in during complete.
```
	GroupLister interfaces.LDAPGroupLister
	GroupMemberExtractor interfaces.LDAPMemberExtractor
	UserNameMapper interfaces.LDAPUserNameMapper
	GroupNameMapper interfaces.LDAPGroupNameMapper
	GroupClient client.GroupInterface
	Host string
```

stevekuznetsov replied:
I'm going to keep *this* struct parallel with those in other commands - the `Options` contains options given as input to the command on the command line.

stevekuznetsov replied:
Now that I finally understand what you mean here, I can respond:
I don't think the above is a good idea. It makes sense to use the `Options` struct to hold, well, CLI options. `Run()` will create the above information as necessary as it creates a `GroupSyncer` and then executes the syncer.

deads2k commented on pkg/cmd/experimental/syncgroups/sync-groups.go:
```diff
@@ -0,0 +1,188 @@
+}
+
+func NewCmdSyncGroups(name, fullName string, f *clientcmd.Factory, out io.Writer) *cobra.Command {
+	options := NewSyncGroupsOptions()
```
Create a different struct for holding the flag values.  Pass that object into `Complete`

deads2k commented on pkg/cmd/experimental/syncgroups/sync-groups.go:
```diff
@@ -0,0 +1,188 @@
+	return cmd
+}
+
+func (o *SyncGroupsOptions) Complete(args []string, f *clientcmd.Factory) error {
```
Have this method do all the config parsing and testing required to build exactly the correct values for the `LDAPGroupSyncer`.  This is the **only** spot in all the code that inspects all the information.  Everything else takes the exact minimum required.

deads2k commented on pkg/auth/authenticator/password/ldappassword/ldaputil/identityfactory.go:
```diff
@@ -0,0 +1,70 @@
+
+// LDAPUserToIdentityMapping holds LDAP attributes that map to OpenShift Identity fields in oder to
+// create a deterministic mapping from LDAP user entries to OpenShift Identities
+type LDAPUserToIdentityMapping struct {
```
You were going to add helpers for `GetEmail` etc

stevekuznetsov replied:
yes

deads2k commented on pkg/auth/authenticator/password/ldappassword/ldap.go:
```diff
@@ -15,41 +12,33 @@ import (
 
 	authapi "github.com/openshift/origin/pkg/auth/api"
 	"github.com/openshift/origin/pkg/auth/authenticator"
+	"github.com/openshift/origin/pkg/auth/authenticator/password/ldappassword/ldaputil"
```
I still want these changes in a separate pull.

stevekuznetsov replied:
#4264

stevekuznetsov commented:
@deads2k Confirming that this is ready for review when you return to the office.

deads2k commented on pkg/cmd/experimental/syncgroups/schema1/newsyncer.go:
```diff
@@ -0,0 +1,66 @@
+
+// NewGroupSyncer creates a GroupSyncer for a schema-specific config written for an LDAP schema with
+// first-class group entries that hold membership lists as attributes
+func NewGroupSyncer(clientConfig ldaputil.ClientConfig,
```
move this to the command.

deads2k replied:
This should be something like `NewSchema1GroupSyncer`.

deads2k commented on pkg/cmd/experimental/syncgroups/schema1/newsyncer.go:
```diff
@@ -0,0 +1,66 @@
+
+	// In order to build the GroupNameMapper, we need to know if the user defined a hard mapping
+	// or one based on LDAP group entry attributes
+	if ldapSyncConfig.CommonConfig.UserDefinedGroupNameMapping {
```
Can this hide in a function `GetGroupMapping(ldapGroupUIDToOpenShiftNameMapping map[string]string, interfaces.LDAPGroupGetter)`?

deads2k replied:
Should this just be `userDefinedGroupNameMapping`?

stevekuznetsov replied:
passing a boolean into a function to make a decision on it first-thing smells bad

stevekuznetsov replied:
also this refactor into a method ends up being just as long in terms of lines of code. make your choice, I'm fairly ambivalent

deads2k commented on pkg/cmd/experimental/syncgroups/schema1/newsyncer.go:
```diff
@@ -0,0 +1,66 @@
+	// Syncing a whitelist of OpenShift groups will require a LocalGroupLister
+	// Syncing all LDAP groups will require us to use the ldapInterface as the lister
+	// Syncing all OpenShift groups will require a AllLocalGroupLister
+	switch scope {
```
This looks like a `GetGroupLister(scope, whitelist, groupInterface, ldapHost)`

deads2k commented on pkg/cmd/experimental/syncgroups/grouplister.go:
```diff
@@ -0,0 +1,103 @@
+)
+
+// NewAllOpenShiftGroupLister returns a new AllLocalGroupLister
+func NewAllOpenShiftGroupLister(host string, groupClient osclient.GroupInterface) interfaces.LDAPGroupLister {
```
Better name of `host`.  `ldapURL`?  Add to the godoc explaining how the `ldapURL` will be used.

deads2k replied:
This and the whitelist version below also provide an implicit: `GroupNameFor` method based on the reverse mapping.

stevekuznetsov replied:
Why is the implicit reverse method useful to expose?

deads2k commented on pkg/cmd/experimental/syncgroups/grouplister.go:
```diff
@@ -0,0 +1,103 @@
+}
+
+func (l *AllLocalGroupLister) ListGroups() (ldapGroupUIDs []string, err error) {
+	operator := labels.EqualsOperator
```
just reference the constant inline.  You aren't paid by the line.

stevekuznetsov replied:
lol

deads2k commented on pkg/cmd/experimental/syncgroups/grouplister.go:
```diff
@@ -0,0 +1,103 @@
+
+func (l *AllLocalGroupLister) ListGroups() (ldapGroupUIDs []string, err error) {
+	operator := labels.EqualsOperator
+	val := util.NewStringSet(l.host)
```
create inline

deads2k commented on pkg/cmd/experimental/syncgroups/grouplister.go:
```diff
@@ -0,0 +1,103 @@
+		return nil, fmt.Errorf("unexpected error gemerating label selector: %v", err)
+	}
+	label := labels.LabelSelector{*requirement}
+	field := fields.Everything()
```
inline.

deads2k commented on pkg/cmd/experimental/syncgroups/util/util.go:
```diff
@@ -0,0 +1,177 @@
+}
+
+// GetOpenShiftGroups uses a client to retrieve all groups from the names given
+func GetOpenShiftGroups(names []string, client client.GroupInterface) ([]ouserapi.Group, error) {
```
move to point of use and don't make public.

deads2k commented on pkg/cmd/experimental/syncgroups/grouplister.go:
```diff
@@ -0,0 +1,103 @@
+}
+
+// NewWhitelistOpenShiftGroupLister returns a new LocalGroupLister
+func NewWhitelistOpenShiftGroupLister(whitelist []string, client osclient.GroupInterface) interfaces.LDAPGroupLister {
```
`openshiftWhitelist` and indicate in godoc that this whitelist is for the openshift groups, not the resulting ldap groupUID

deads2k commented on pkg/cmd/experimental/syncgroups/groupnamemapper.go:
```diff
@@ -0,0 +1,58 @@
+)
+
+// NewUserDefinedGroupNameMapper returns a new UserDefinedLDAPGroupNameMapper
+func NewUserDefinedGroupNameMapper(mapping map[string]string) interfaces.LDAPGroupNameMapper {
```
explain what mapping is.  I think it's ldapGroupUID to openshiftGroupName

deads2k commented on pkg/cmd/experimental/syncgroups/groupnamemapper.go:
```diff
@@ -0,0 +1,58 @@
+	nameMapping map[string]string
+}
+
+func (m *UserDefinedLDAPGroupNameMapper) GroupNameFor(ldapGroupUID string) (openShiftGroupName string, err error) {
```
don't name the return values.

deads2k commented on pkg/cmd/experimental/syncgroups/groupnamemapper.go:
```diff
@@ -0,0 +1,58 @@
+
+// EntryAttributeLDAPGroupNameMapper references the name attribute mapping to determine which attribute
+// of a first-class LDAP group entry should be used as the OpenShift Group name for the resource
+type EntryAttributeLDAPGroupNameMapper struct {
```
This structure feels wrong, but I don't feel like leaning on it now.  I think I'll tolerate it.

deads2k commented on pkg/cmd/experimental/syncgroups/groupsyncer.go:
```diff
@@ -0,0 +1,132 @@
+
+	// add LDAP-sync-specific annotations
+	group.Annotations[syncutil.LDAPUID] = ldapGroupUID
+	group.Annotations[syncutil.LDAPSyncTime] = strconv.FormatInt(time.Now().Unix(), 10)
```
Let's assume our users don't speak "seconds since epoch".  ISO8601?

stevekuznetsov replied:
Is there a utility function for this somewhere?

deads2k replied:
> Is there a utility function for this somewhere?

Probably.  I leave you to google-fu.

deads2k commented on pkg/cmd/experimental/syncgroups/interfaces/interfaces.go:
```diff
@@ -0,0 +1,37 @@
+}
+
+// LDAPUserGetter maps a ldapUserUID to a first-class LDAP user entry
+type LDAPUserGetter interface {
```
Looks dead.  Remove for now.

stevekuznetsov replied:
https://github.com/openshift/origin/pull/3996/files#diff-5f5b97b214111381809fa81d6396a989R81

deads2k replied:
> https://github.com/openshift/origin/pull/3996/files#diff-5f5b97b214111381809fa81d6396a989R81

doesn't rely on this interface, only the method.  Kill it.

deads2k commented on pkg/cmd/experimental/syncgroups/sync-groups.go:
```diff
@@ -0,0 +1,308 @@
+  $ %[1]s ldap all --sync-config=/path/to/ldap-sync-config.yaml
+
+  // Sync specific groups specified in a whitelist file with an LDAP server 
+  $ %[1]s ldap whitelist --whitelist-source=/path/to/whitelist.txt --sync-config=/path/to/sync-config.yaml
```
I definitely don't like repeating myself with `whitelist --whitelist-source`.  Additional arguments implies `whitelist`.  `--whitelist-source` should should be `--whitelist`.  `all` should be the default unless `whitelist` arguments or flag is present.

deads2k replied:
`ldap` should be the default.  `openshift` should be a boolean flag that is false by default.

stevekuznetsov replied:
While these seem like sane defaults I feel like they greatly hinder UX as now each of the four choices is a "special case." In my current scheme, sure, `whitelist-source` may be redundant, but at least each and ever incarnation of `sync-groups` accepts the same form - `source scope args`. @liggitt @smarterclayton does that matter at all?

stevekuznetsov replied:
Also furthermore if we want to allow for OpenShift group whitelist files and LDAP group uid whitelist literals, we need to support

```sh
$ oadm sync-groups ldapGroupUID1 --sync-config=... #sync whitelist literal from LDAP
```

This is still possible, obviously, but then you need to do validation right away to determine if the first argument is a search source or a group UID... seems messy.

deads2k commented on pkg/cmd/experimental/syncgroups/sync-groups.go:
```diff
@@ -0,0 +1,308 @@
+
+// readLines interprets a file as plaintext and returns a string array of the lines of text in the file
+func readLines(string path) ([]string, error) {
+	file, err := os.Open(path)
```
`ioutil.ReadFile` and `strings.Split` ought to do.

stevekuznetsov replied:
Do we want to worry about lines ending in a carriage return versus a new line?

deads2k commented on pkg/cmd/experimental/syncgroups/sync-groups.go:
```diff
@@ -0,0 +1,308 @@
+				o.WhitelistContents,
+			)
+		}
+		if serverapi.IsLDAPSpecificQueryFromAttributeTemplate(config.GroupQueryTemplate) {
```
kill this for now.

deads2k commented on pkg/cmd/experimental/syncgroups/sync-groups.go:
```diff
@@ -0,0 +1,308 @@
+	//   - An LDAP server with first-class groups and users where group membership is held in user attributes
+	//     will have a specific query from attribute template for groups
+	if config.GroupQueryTemplate == nil {
+		// schema 2
```
eliminate the schema2 and schema3 bits for now

deads2k commented on pkg/cmd/experimental/syncgroups/util/util.go:
```diff
@@ -0,0 +1,177 @@
+}
+
+// GetKeys returns a string array of the keys of any map[string][]string
+func GetKeys(theMap map[string][]string) (keys []string) {
```
https://github.com/openshift/origin/blob/master/Godeps/_workspace/src/k8s.io/kubernetes/pkg/util/set.go#L40

stevekuznetsov replied:
Thanks, I knew it was somewhere but I couldn't find it.

deads2k commented on pkg/cmd/server/api/types.go:
```diff
@@ -646,3 +646,75 @@ type PodManifestConfig struct {
+type LDAPSyncConfig struct {
+	// URL is a partial RFC 2255 URL which specifices the scheme, host and port of the LDAP server
+	// The syntax of the URL is: `scheme://host:port/`, where `scheme` is either ldap or ldaps
+	URL string
```
there ought to be a common type to describe `URL`, `Insecure`, and `CA`.

deads2k replied:
Looks like you're missing a couple options from `LDAPPasswordIdentityProvider` for making the connection.

stevekuznetsov replied:
This is a good idea. I'll combine those two somehow

deads2k commented on pkg/cmd/server/api/types.go:
```diff
@@ -646,3 +646,75 @@ type PodManifestConfig struct {
+	CA string
+
+	// Attributes maps LDAP attributes to identities
+	Attributes LDAPAttributes
```
`LDAPAttributes` needs a better name.  So does `Attributes`

deads2k commented on pkg/cmd/server/api/types.go:
```diff
@@ -646,3 +646,75 @@ type PodManifestConfig struct {
+
+	// UserDefinedGroupNameMapping determines if the mapping of LDAP group UIDs to OpenShift Group
+	// names is defined by an input mapping or algorithmically
+	UserDefinedGroupNameMapping bool
```
Extraneous.  If the map is empty, then don't use it.  If it has values, then use it.

deads2k commented on pkg/cmd/server/api/types.go:
```diff
@@ -646,3 +646,75 @@ type PodManifestConfig struct {
+	UserQueryTemplate LDAPQueryTemplate
+}
+
+type LDAPQueryTemplate struct {
```
I'm not convinced this type is correct.  Why are there multiple types representing this above?

stevekuznetsov replied:
Depending on the schema, the type of template (are we grabbing something specific? do we need to insert something from the attribute field of another LDAP entry) will change. I couldn't figure out how to unite the schema-specific query templates for groups and users without this ugly superset impl. Suggestions welcome.

stevekuznetsov replied:
I will refactor this to use `RawExtension`

deads2k commented on pkg/cmd/server/api/types.go:
```diff
@@ -646,3 +646,75 @@ type PodManifestConfig struct {
+	LDAPGroupUIDToOpenShiftNameMapping map[string]string
+
+	// GroupQueryTemplate holds the (optional) template for an LDAP query that returns group entries
+	GroupQueryTemplate LDAPQueryTemplate
```
This structure doesn't seem right.  I'm probably willing to fight over it later.

deads2k commented on pkg/cmd/server/api/validation/validation.go:
```diff
@@ -224,3 +225,51 @@ func ValidateExtendedArguments(config api.ExtendedArguments, flagFunc func(*pfla
+	var errors fielderrors.ValidationErrorList
+	parsedURL, err := url.Parse(config.URL)
+	if err != nil {
+		errors = append(errors, fmt.Errorf("given LDAP url invalid: %v", err))
```
use the `fielderrors` like the rest of validation.

deads2k commented on pkg/cmd/server/api/validation/validation.go:
```diff
@@ -224,3 +225,51 @@ func ValidateExtendedArguments(config api.ExtendedArguments, flagFunc func(*pfla
+		}
+	}
+
+	templates := []api.LDAPQueryTemplate{config.GroupQueryTemplate, config.UserQueryTemplate}
```
create a method to check `LDAPQueryTemplate` types and then use `.Prefix()` to get the field names correct.

deads2k commented on test/integration/syncgroups_ldap_test.go:
```diff
@@ -0,0 +1,540 @@
+	testutil.RequireDocker()
+}
+
+func TestLDAPSync(t *testing.T) {
```
how long does it take this test to run?

stevekuznetsov replied:
Deprecated, will remove.

stevekuznetsov commented:
@deads2k Still working on some things, will tag when ready for review.

stevekuznetsov commented:
@deads2k Ready for review sans `RawExtension`.

If we want to use `RawExtension` (which we do) and we want the group query to be optional (which we do, necessary for schema 2), we run into an issue where there's some sort of `<nil>` object behind the `EmbeddedObject` and the swtich/case on it's type won't be pretty. Thoughts?

liggitt commented:
@deads2k after living with both auth config and various bits of the API, do you prefer a single RawExtension/EmbeddedObject field or an embedded struct with pointer fields to specific types, only one of which is non-nil? It's a switch either way, one gets you unique field names, which I sort of like, and removes kind/version clutter from the serialized version

deads2k commented:
> @deads2k after living with both auth config and various bits of the API, do you prefer a single RawExtension/EmbeddedObject field or an embedded struct with pointer fields to specific types, only one of which is non-nil? It's a switch either way, one gets you unique field names, which I sort of like, and removes kind/version clutter from the serialized version

I really dislike the "multiple pointers only one of which non-nil" approach.  I think I dislike it more than I dislike the "EmbeddedObject that only accepts a subset of the available objects" approach.  When I look through how the build code navigates branches, I think the RawExtension type switch comes out cleaner and that builds an object that reflects the "exactly one"-ness of the structure.

stevekuznetsov commented:
@deads2k Implemented the approach you suggested. Ready for review.

deads2k commented on pkg/cmd/experimental/syncgroups/grouplister.go:
```diff
@@ -0,0 +1,123 @@
+
+// AllLocalGroupLister lists unique identifiers for LDAP lookup of all local OpenShift Groups that
+// have been marked with an LDAP URL annotation as a result of a previous sync.
+type AllLocalGroupLister struct {
```
Wasn't this going to be a `GroupNameMapper` too, so that you could be certain that when you resync'ed, the same openshift group was updated?

stevekuznetsov replied:
I don't think it was. If you don't change your config, you will be deterministically mapping these groups to their names anyway. The  dynamic `GroupNameMapper` uses a `GroupGetter` (which you commented on looking bad earlier). This `GroupGetter` necessarily uses a `GroupLister` to gain information about groups in the first place, which brings us back to this object. It does in the end guarantee that you sync the same group to the same place twice.

deads2k replied:
Open an issue for it.  I don't feel like blocking this pull on it, but since openshift group names are built based on ldap metadata, unless this forces a name mapping, syncing this group could create a new group altogether and not update this at all.  I want avoid that condition.

stevekuznetsov replied:
This does force a mapping. Nothing is forcing *you* as the admin to keep the mapping the same between invocations... but if you do, nothing will change.

deads2k commented on pkg/cmd/experimental/syncgroups/groupsyncer.go:
```diff
@@ -0,0 +1,143 @@
+// ISO8601 returns an ISO 6801 formatted string from a time.
+func ISO8601(t time.Time) string {
+	var tz string
+	if zone, offset := t.Zone(); zone == "UTC" {
```
Always use "Z", you'll be a happier person.

stevekuznetsov replied:
Given that the code implements the standard correctly, what would you like me to change? From Zulu to `0` offset?

deads2k replied:
> Given that the code implements the standard correctly, what would you like me to change? From Zulu to 0 offset?

Using the current timezone always confuses people when they start dealing with daylight saving time.  I'm suggesting that you normalize everything to UTC, so that no one writes issues about how 4am is ordered after 4:30am.  I suppose that you should probably be consistent with metadata.creationTime.  See what that does.

deads2k commented on pkg/cmd/experimental/syncgroups/schema1/ldapinterface.go:
```diff
@@ -0,0 +1,184 @@
+// - LDAPMemberExtractor
+// - LDAPGroupGetter
+// - LDAPGroupLister
+type LDAPInterface struct {
```
Note that this is not thread-safe.  Otherwise, someone may try it.

deads2k commented on pkg/cmd/experimental/syncgroups/sync-groups.go:
```diff
@@ -0,0 +1,356 @@
+  $ %[1]s --whitelist=/path/to/whitelist.txt --sync-config=/path/to/sync-config.yaml
+
+  // Sync all OpenShift Groups that have already been synced with an LDAP server
+  $ %[1]s openshift --sync-config=/path/to/ldap-sync-config.yaml
```
instead of an `openshift` arg, maybe `--existing-groups` or some other flag indicating that we're resyncing what's already sync'ed, but nothing new.

deads2k commented on pkg/cmd/experimental/syncgroups/sync-groups.go:
```diff
@@ -0,0 +1,356 @@
+	}
+
+	if len(o.WhitelistSource) != 0 {
+		whitelistData, err := readLines(o.WhitelistSource)
```
Trim these.  Someone will put whitespace in there to make life difficult.

deads2k commented on pkg/cmd/experimental/syncgroups/util/util.go:
```diff
@@ -0,0 +1,122 @@
+package config
```
util packages are usually a dumping ground.  Do you have a better name?  Oh, and make your package name match the folder name.

deads2k commented on pkg/cmd/experimental/syncgroups/util/util.go:
```diff
@@ -0,0 +1,122 @@
+// QueryForUniqueEntry queries for an LDAP entry on an LDAP server determined from a ClientConfig
+// by creating a search request from the requisite input. The query is expected to return one unqiue
+// result. If this is not the case, errors are raised
+func QueryForUniqueEntry(clientConfig ldaputil.LDAPClientConfig,
```
Does this belong with the `ldaputil` stuff?

deads2k commented on pkg/cmd/experimental/syncgroups/util/util.go:
```diff
@@ -0,0 +1,122 @@
+	return entries, nil
+}
+
+func CreateClientConfig(URL string, insecure bool, CA string) (ldaputil.LDAPClientConfig, error) {
```
Move to `ldaputil`

deads2k commented on pkg/cmd/server/api/types.go:
```diff
@@ -699,3 +699,93 @@ type AssetExtensionsConfig struct {
+	LDAPClientInfo LDAPPasswordIdentityProvider
+
+	// GroupNameAttributes (optionally) maps LDAP group entries to OpenShift Groups
+	GroupNameAttributes []string
```
isn't this specific to schema1 or schema3?

stevekuznetsov replied:
@deads2k @liggitt I need your input on this - as it stands I chose not to include this optional and schema-specific field in the QueryTemplates because I really liked how that was a non-negotiable part of the sync config - regardless of what schema you have, the query templates will provide the query templates. I didn't quite like to have a catch-all "schema-specific-config" object to shove something like this in. Should it remain as an optional field here or should there be another `EmbeddedObject` to hold other schema-specific config? Should there be just one catch-all schema-specific config object?

deads2k replied:
> Should it remain as an optional field here or should there be another EmbeddedObject to hold other schema-specific config? Should there be just one catch-all schema-specific config object?

Schema specific config options should be contained in the schema specific config objects.  I don't think having a congruent name is worth having sometimes valid config options.  I also don't have an issue with including this in an object called `Schema1QueryTemplates`, so I don't think you lose your nice name.

deads2k commented:
Separate the registration of your command and the external config type packages into a separate pull.  That will still allow us accept most of this work and test it in extended tests without exposing it.

stevekuznetsov commented:
@deads2k Broke things apart. If you like the breaks, I can pull the command out of this PR. Would like comments on impl of `Run` method for `Objects`, comments on possible changes to API objects.

danmcp commented:
@deads2k Bump

stevekuznetsov commented:
@liggitt 
While I build the test for this, I will keep the last commit in to show you how a syncer is built. Please review for LDAP specifics. I will fix the API objects as soon as I get `go vet` sorted.

deads2k commented:
> @deads2k Bump

This is waiting on 3.0.2 to be officially cut (to avoid muddying the commands) and for @liggitt to finish a review on the ldap queries themselves.  I can make sure they're not insane, but he's more familiar with that syntax than I am.

deads2k commented on pkg/cmd/openshift/openshift.go:
```diff
@@ -160,6 +161,7 @@ func newExperimentalCommand(name, fullName string) *cobra.Command {
 	experimental.AddCommand(exipfailover.NewCmdIPFailoverConfig(f, fullName, "ipfailover", out))
 	experimental.AddCommand(buildchain.NewCmdBuildChain(name, fullName+" "+buildchain.BuildChainRecommendedCommandName, f, out))
 	experimental.AddCommand(diagnostics.NewCommandDiagnostics("diagnostics", fullName+" diagnostics", out))
+	experimental.AddCommand(syncgroups.NewCmdSyncGroups(syncgroups.SyncGroupsRecommendedName, fullName+" "+syncgroups.SyncGroupsRecommendedName, f, out))
```
Remove this line from your pull.  Removing this line will prevent the command from being registered.

deads2k commented:
Remove the one line I called out.  Then build an extended test to test this functionality.  You should be able to manually build the sync job and prove that this works.

deads2k commented:
[test] [extended:core]

deads2k commented:
Ah, I see.  Test doesn't compile yet.

deads2k commented on pkg/cmd/experimental/syncgroups/schema1/ldapinterface.go:
```diff
@@ -0,0 +1,183 @@
+}
+
+// ExtractMembers returns the LDAP member entries for a group specified with a ldapGroupUID
+func (e *LDAPInterface) ExtractMembers(ldapGroupUID string) (members []*ldap.Entry, err error) {
```
Don't name the return args in the implementations.  Its good to do in the interface definitions, but not the impls

deads2k commented on pkg/cmd/experimental/syncgroups/sync-groups.go:
```diff
@@ -0,0 +1,335 @@
+package syncgroups
```
underscores, not dashes in filename.

deads2k commented on pkg/cmd/experimental/syncgroups/sync-groups.go:
```diff
@@ -0,0 +1,335 @@
+
+// getGroupNameMapper returns an LDAPGroupNameMapper either by using a user-provided mapping or
+// by creating a mapper that uses an algorithmic mapping
+func getGroupNameMapper(userDefinedMapping map[string]string,
```
Change this signature to take a `defaultGroupNameMapper` in place of the `groupGetter`.

stevekuznetsov replied:
I don't understand this comment. What is a `defaultGroupNameMapper` ? When should it be chosen (to be regurgitated by the method?) instead of constructing new mappers from the config?

liggitt commented on pkg/cmd/experimental/syncgroups/schema1/ldapinterface.go:
```diff
@@ -0,0 +1,183 @@
+
+var _ interfaces.LDAPMemberExtractor = &LDAPInterface{}
+var _ interfaces.LDAPGroupGetter = &LDAPInterface{}
+var _ interfaces.LDAPGroupLister = &LDAPInterface{}
```
move to the test file

liggitt commented on pkg/cmd/experimental/syncgroups/schema1/ldapinterface.go:
```diff
@@ -0,0 +1,183 @@
+	attributes.Insert(e.userQueryTemplate.SpecificFilterSourceAttribute)
+
+	// build specific filter by injecting ldapGroupUID
+	specificFilter := fmt.Sprintf(e.groupQueryTemplate.SpecificFilterBase, ldapGroupUID)
```
What if ldapGroupUid is a DN? How would they construct the SpecificFilterBase to query for it

liggitt replied:
EscapeFilter ldapGroupUID here and plug in the result verbatim below

liggitt replied:
also, validate the SpecificFilterBase using ldap.CompileFilter(filter) (may need to replace the %s with a dummy value before validating)

liggitt replied:
should this follow the user lookup pattern for naming?
baseDN - defines tree we care about
filter - limits to all groups we care about
queryAttribute - name of the attribute the ldapGroupUid matches

liggitt replied:
special case the "dn" query attribute

liggitt commented on pkg/cmd/experimental/syncgroups/schema1/ldapinterface.go:
```diff
@@ -0,0 +1,183 @@
+	specificFilter := fmt.Sprintf(e.groupQueryTemplate.SpecificFilterBase, ldapGroupUID)
+	// build filter by conjoining specific filter and common filter
+	fullFilter := fmt.Sprintf("(&(%s)(%s))",
+		ldap.EscapeFilter(e.groupQueryTemplate.Options.CommonFilter),
```
don't escape the CommonFilter, it should already be in a valid filter format

liggitt commented on pkg/cmd/experimental/syncgroups/schema1/ldapinterface.go:
```diff
@@ -0,0 +1,183 @@
+// determining which attributes to search for with a LDAPuserAttributeDefiner
+func (e *LDAPInterface) queryForUser(ldapUserUID string) (user *ldap.Entry, err error) {
+	// build specific filter by injecting ldapUserUID
+	specificFilter := fmt.Sprintf(e.userQueryTemplate.Template.SpecificFilterBase, ldapUserUID)
```
same escaping comments, refactor to treat consistently with group query

stevekuznetsov commented:
@deads2k @liggitt Addressed comments, ready for review. Learned that it is not possible to create a label selector for a `host:port` value, so we need to re-think how we label groups as far as the server from which they were synced.

deads2k commented:
> @deads2k @liggitt Addressed comments, ready for review. Learned that it is not possible to create a label selector for a host:port value, so we need to re-think how we label groups as far as the server from which they were synced.

This isn't exposed yet.  Make it an annotation for now and we'll revisit.

deads2k commented:
Also, it didn't compile

deads2k commented:
gofmt

deads2k commented:
Also, no longer a WIP

deads2k commented on pkg/cmd/experimental/syncgroups/rfc2307/ldapinterface.go:
```diff
@@ -0,0 +1,154 @@
+// - LDAPMemberExtractor
+// - LDAPGroupGetter
+// - LDAPGroupLister
+type LDAPInterface struct {
```
Where' the `_test` file with type assertions?

deads2k commented on pkg/cmd/experimental/syncgroups/rfc2307/ldapinterface.go:
```diff
@@ -0,0 +1,154 @@
+	// userQuery holds the information necessary to make an LDAP query for a specific
+	// first-class user entry on the LDAP server
+	userQuery ldaputil.IdentifiyingLDAPQueryOptions
+	// groupMembershipAttributes defines which attributes on an LDAP user entry will be interpreted
```
This comment isn't what I was expecting.  It suggests that the LDAP user entry has attributes that describe group membership. Wasn't this schema "first class groups, first class users, group membership on the groups"?  Also, clarify that in your comment on the type.

stevekuznetsov replied:
Copy-paste user error. My apologies.

deads2k commented:
put updates for comments in a separate commit.

deads2k commented on pkg/auth/ldaputil/query.go:
```diff
@@ -0,0 +1,211 @@
+// buildDNQuery builds the query that finds an LDAP entry with the given DN
+// this is done by setting the DN to be the base DN for the search and setting the search scope
+// to only consider the base object found
+func (o *IdentifiyingLDAPQueryOptions) buildDNQuery(dn string,
```
@liggitt is this sane?

deads2k commented on pkg/auth/ldaputil/query.go:
```diff
@@ -0,0 +1,211 @@
+
+// NewSearchRequest creates a new search request from the identifying query by internalizing the value of
+// the attribute to be filtered as well as any additional attributest that need to be recovereds
+func (o *IdentifiyingLDAPQueryOptions) NewSearchRequest(attributeValue string,
```
Add unit tests for this guy.

deads2k commented on pkg/auth/ldaputil/query.go:
```diff
@@ -0,0 +1,211 @@
+	allAttributes := util.NewStringSet(o.NameAttributes...)
+	allAttributes.Insert(additionalAttributes...)
+
+	if o.QueryAttribute == "DN" || o.QueryAttribute == "dn" {
```
why not case insensitive?

deads2k commented on pkg/auth/ldaputil/query.go:
```diff
@@ -0,0 +1,211 @@
+
+	if o.QueryAttribute == "DN" || o.QueryAttribute == "dn" {
+		if !strings.Contains(attributeValue, o.BaseDN) {
+			return nil, fmt.Errorf("search for entry with %s=%s would search outside of the base dn specified (dn=%s)",
```
Create a particular error type for this.  See https://github.com/kubernetes/kubernetes/blob/master/pkg/client/unversioned/clientcmd/validation.go#L32 as an example

deads2k commented on pkg/auth/ldaputil/query.go:
```diff
@@ -0,0 +1,211 @@
+
+	// NameAttributes defines the attributes for the LDAP entries returned by the Query that will be interpreted
+	// as their names
+	NameAttributes []string
```
I'd prefer to have this as a param to the `NewSearchRequest` function.

stevekuznetsov replied:
If we do this, there's no need for this type, everything can use `LDAPQueryOnAttribute` and hold the `NameAttributes` separately. I like this because it nicely bundles up almost everything you need for creating search requests, but I don't have strong feelings about it.

deads2k commented on pkg/auth/ldaputil/query.go:
```diff
@@ -0,0 +1,211 @@
+		return nil, err
+	}
+	if len(result) > 1 {
+		return nil, fmt.Errorf("multiple entries found matching %s", query.Filter)
```
Add the DNs to this message.

deads2k replied:
> Add the DNs to this message.

Add the number of entries and the first 10 dns.

deads2k commented on pkg/auth/ldaputil/query.go:
```diff
@@ -0,0 +1,211 @@
+	// No entries returned from the LDAP search request means that the LDAP search was not configured
+	// correctly. The search must return with at least one LDAP entry
+	if len(entries) == 0 {
+		return nil, fmt.Errorf("no LDAP user entry found for filter: %s", query.Filter)
```
not necessarily a user

deads2k replied:
Don't make this error on empty.  You're issuing a search.  The search succeeded, you simply didn't get a hit.

stevekuznetsov replied:
I believe the reason that this errors on empty is that when you are looking for a specific LDAP entry that you know exists (as you have the UID for it), the behavior of the LDAP sync would *not* be as expected if the entry was not found. For example, if you want to sync a group and screwed up the user filter such that the users of the group are not a subset of all users from that filter, the sync will effectively do nothing, and you'd expect as a user to get an error instead of silently doing nothing.

deads2k replied:
> I believe the reason that this errors on empty is that when you are looking for a specific LDAP entry that you know exists (as you have the UID for it), the behavior of the LDAP sync would not be as expected if the entry was not found. For example, if you want to sync a group and screwed up the user filter such that the users of the group are not a subset of all users from that filter, the sync will effectively do nothing, and you'd expect as a user to get an error instead of silently doing nothing.

Find usages and fix it.  A `List` that returns `{}` did not fail.

deads2k commented on pkg/auth/ldaputil/query.go:
```diff
@@ -0,0 +1,211 @@
+	if err != nil {
+		return nil, err
+	}
+	if len(result) > 1 {
```
`!= 1` since `QueryForEntries` shouldn't error on zero hits.

deads2k commented on pkg/auth/ldaputil/query.go:
```diff
@@ -0,0 +1,211 @@
+// QueryForUniqueEntry queries for an LDAP entry on an LDAP server determined from a ClientConfig
+// by creating a search request from the requisite input. The query is expected to return one unqiue
+// result. If this is not the case, errors are raised
+func QueryForUniqueEntry(clientConfig LDAPClientConfig,
```
screens are wide, put this declaration on one line.  Its annoying me to see multiple lines everywhere.

deads2k replied:
Either return a typed error for `NotFound` or allow returning `nil` if no hits are found.  I don't really care which.

stevekuznetsov replied:
I'm on a 1080p screen and have two windows open side-by-side. 110 columns scrapes the side for me.

deads2k commented on pkg/auth/ldaputil/query.go:
```diff
@@ -0,0 +1,211 @@
+// by creating a search request from the requisite input. The query is expected to return one unqiue
+// result. If this is not the case, errors are raised
+func QueryForUniqueEntry(clientConfig LDAPClientConfig,
+	query *ldap.SearchRequest) (entry *ldap.Entry, err error) {
```
Don't name return values for function implementations.  Find them all in this pull and remove them.

deads2k commented on pkg/cmd/openshift/openshift.go:
```diff
@@ -160,6 +161,8 @@ func newExperimentalCommand(name, fullName string) *cobra.Command {
 	experimental.AddCommand(buildchain.NewCmdBuildChain(name, fullName+" "+buildchain.BuildChainRecommendedCommandName, f, out))
 	experimental.AddCommand(diagnostics.NewCommandDiagnostics("diagnostics", fullName+" diagnostics", out))
+	// TOOD(skuznets): Re-enable this command once ready
+	experimental.AddCommand(syncgroups.NewCmdSyncGroups(syncgroups.SyncGroupsRecommendedName, fullName+" "+syncgroups.SyncGroupsRecommendedName, f, out))
```
@liggitt I'm ok with adding this as experimental now that 3.0.2 is done.  Complaints?

stevekuznetsov replied:
Can revert, added so that tree was build correctly for compilation (nowhere else called this code)

deads2k commented on pkg/cmd/server/api/v1/types.go:
```diff
@@ -696,3 +697,128 @@ type AssetExtensionsConfig struct {
 	HTML5Mode bool `json:"html5Mode"`
 }
+
+type LDAPSyncConfig struct {
```
@liggitt This type isn't registered externally, so you can't use it yet.  I'd rather not argue about it in this pull.  It needs reworking, but lets rework it elsewhere.

deads2k commented on pkg/cmd/server/api/v1/types_test.go:
```diff
@@ -1,21 +1,21 @@
 
 import (
-	"testing"
+  "testing"
```
I expect no changes in this file

stevekuznetsov replied:
GoTools failing me here... sorry

deads2k commented on pkg/cmd/experimental/syncgroups/groupsyncer.go:
```diff
@@ -0,0 +1,141 @@
+}
+
+// LDAPGroupSyncer sync Groups with records on an external LDAP server
+type LDAPGroupSyncer struct {
```
Unit tests for this type.  Stub in test structs for the values and make sure the flows are correct and errors are handled properly.

stevekuznetsov commented:
@deads2k Addressed comments. Added unit tests. Extended test is failing to connect to the LDAP server, and I cannot get the build for the server to ever leave the first stage, having issues pulling an image from the hub or something? Unclear.

stevekuznetsov commented:
[test][extended]

stevekuznetsov commented:
Flake on Travis in [`test-cmd`](https://travis-ci.org/openshift/origin/jobs/82262917#L630):
```
!!! Error in test/cmd/admin.sh:76
	'oadm policy add-cluster-role-to-user cluster-admin system:no-user' exited with status 1
Call stack:
	1: test/cmd/admin.sh:76 main(...)
Exiting with status 1
!!! Error in hack/test-cmd.sh:269
	'${test}' exited with status 1
Call stack:
	1: hack/test-cmd.sh:269 main(...)
Exiting with status 1
[FAIL] !!!!! Test Failed !!!!
```

cc: @liggitt

deads2k commented:
@stevekuznetsov squash this to something more reasonable.  I'd guess
 1. Sync implementation
 1. Command
 1. Config objects

Fix the problem with "I will not merge this. That contract is insane. Lists don't fail on empty result sets." and I think I'll merge it warts and all unless @liggitt yells.
